File: 380. Insert Delete GetRandom O(1)/idg.cpp
#include <bits/stdc++.h>
using namespace std;

class RandomizedSet {
private:
    vector<int> nums;                 // stores elements
    unordered_map<int, int> pos;      // value -> index

public:
    RandomizedSet() {}

    bool insert(int val) {
        if (pos.count(val)) return false;

        nums.push_back(val);
        pos[val] = nums.size() - 1;
        return true;
    }

    bool remove(int val) {
        if (!pos.count(val)) return false;

        int index = pos[val];
        int lastElement = nums.back();

        // swap with last element
        nums[index] = lastElement;
        pos[lastElement] = index;

        // remove last
        nums.pop_back();
        pos.erase(val);

        return true;
    }

    int getRandom() {
        if (nums.empty()) return -1;
        int r = rand() % nums.size();
        return nums[r];
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    RandomizedSet rs;

    int n;
    cin >> n;
    cin.ignore(); // consume newline

    while (n--) {
        string op;
        cin >> op;

        if (op == "insert") {
            int x;
            cin >> x;
            cout << (rs.insert(x) ? "true" : "false") << "\n";
        }
        else if (op == "remove") {
            int x;
            cin >> x;
            cout << (rs.remove(x) ? "true" : "false") << "\n";
        }
        else if (op == "getRandom") {
            cout << rs.getRandom() << "\n";
        }
    }

    return 0;
}

File: 380. Insert Delete GetRandom O(1)/idg.kt

import kotlin.random.Random

class RandomizedSet(){

private val list = ArrayList<Int>()
private val map=HashMap<Int,Int>()

fun insert(`val`: Int): Boolean{
    if(map.containsKey(`val`))return false

    list.add(`val`)
    map[`val`]=list.size-1
    return true
}

fun remove(`val`: Int): Boolean{
    if(!map.containsKey(`val`)) return false

    val index=map[`val`]!!
    val lastElement = list.last()

    //swap with last
    list[index]= lastElement
    map[lastElement] = index

    //remove last
    list.removeAt(list.size-1)
    map.remove(`val`)
    return true

}

fun getRandom(): Int{
    return list.random()
}
}


fun main() {

    val rs = RandomizedSet()
    val n = readln().trim().toInt()

    repeat(n) {
        val parts = readln().trim().split(" ")

        when (parts[0]) {

            "insert" -> {
                val value = parts[1].toInt()
                println(rs.insert(value))
            }

            "remove" -> {
                val value = parts[1].toInt()
                println(rs.remove(value))
            }

            "getRandom" -> {
                println(rs.getRandom())
            }
        }
    }
}
